### PR TITLE
Improve screenshot performance

### PR DESCRIPTION
**Pull request by smcclure17**
addresses #32 

Decreases the time taken to take a snapshot from ~6.5 to ~3 seconds  (not including cold starts)

### CODE REVIEW

File: functions/src/index.ts [hunk 1]
```diff
@@ -103,11 +111,22 @@ app.get("/go/:id", (req, res) => {
...
             <meta property="twitter:title" content="${title}"/>
             <meta property="twitter:description" content="${description}"/>
-            <meta property="twitter:image" content="${image}"/>
+            ${
+              image &&
+              `<meta property="og:image" content="${image}" />
+              <meta property="twitter:image" content="${image}"/>`
+            }
+            ${
+              imgHeight &&
```
Comment (smcclure17) [inline functions/src/index.ts:125]: adding image widths and heights makes it easier for the web crawlers to display the preview images (though I'm not 100% sure why)

File: functions/src/index.ts [hunk 2]
```diff
@@ -50,12 +55,13 @@ app.post("/registerUrl", isAPIKeyAuthorized, (req, res) => {
   if (!isValidUrl(req.body.url)) {
     sendAndThrowInvalidUrlError(res, req.body.url);
   }
-  // TODO: Better way to handle missing data than coercing to empty strings?
   const data: ShareLinkFields = {
     url: req.body.url,
-    imageUrl: req.body.imageUrl ?? "",
-    title: req.body.title ?? "",
-    description: req.body.description ?? "",
+    imageUrl: req.body.imageUrl,
```
Comment (smcclure17) [inline functions/src/index.ts:62]: We should really be sanitizing these inputs--I'll address this in a new PR

File: functions/src/index.ts [hunk 3]
```diff
@@ -24,13 +25,17 @@ import {
...
 const apiKeyHandler = new APIKeyHandler(firestoreDb);
 const app = express();
 app.use(cors({ origin: "*" }));
+app.use(compression());
+// TODO: Add minInstances: 1 to prevent/limit cold starts once this is used in production.
+// See: https://firebase.google.com/docs/functions/manage-functions#min-max-instances 
 const runtimeOpts = {
   timeoutSeconds: 90,
-  memory: "1GB" as "1GB",
+  memory: "2GB" as "2GB",
```
Comment (smcclure17) [inline functions/src/index.ts]: This feels a little wasteful, but I saw a decent bump in performance (1 to 1.5 seconds saved) with this increase

File: README.md [hunk 4]
```diff
@@ -50,6 +50,8 @@ Requires a valid API key. If you do not have a key and would like to register on
 | `title`   | `string`        | Title meta tag                 | `false` |
 | `description`   | `string`        | Description meta tag  | `false` |
 | `imageUrl`   | `string`        | Url of image to use as image meta tag | `false` |
+| `imageHeight`   | `number`        | Height of preview image (imageUrl) in pixels. | `false` |
+| `imageHeight`   | `number`        | Width of preview image (imageUrl) in pixels. | `false` |
```
Comment (mikelehen) [inline README.md]: Is the default really `false`? What does that mean?

File: functions/src/index.ts [hunk 5]
```diff
@@ -24,13 +25,17 @@ import {
 } from "./utils";
 
 const firestoreDb = firebaseApp.firestore();
+firestoreDb.settings({ ignoreUndefinedProperties: true });
```
Comment (mikelehen) [inline functions/src/index.ts:30]: Maybe add a comment on why we are enabling this?  Like "We have a number of optional fields associated with share links. We want Firestore to just omit them if they come in as undefined." or something.

File: functions/src/index.ts (hunk 3 above)
Comment (mikelehen) [reply functions/src/index.ts]: Maybe add a comment that we're doing this for CPU reasons not memory reasons? 🤷 

Also, I think you can get rid of `as "2GB"` if you do `const runtimeOpts: functions.RuntimeOptions = { ...`

The problem right now is that `runtimeOpts` has an inferred type of `{ timeoutSeconds: number, memory: string }` which is incompatible with the more specific `functions.RuntimeOptions` type which requires `memory` to be one of the specifically allowed string values.  If we give `runtimeOpts` an explicit type up front, then that problem goes away...

File: functions/src/index.ts [hunk 6]
```diff
@@ -145,9 +164,11 @@ app.get("/screenshot", (req, res) => {
   takeScreenshot(screenshotUrl, "temp")
     .then((file: string) => {
       console.log("screenshot generated.");
-      // Let the CDN and the browser cache for 24hrs.
-      console.log("Setting cache-control header.");
-      res.header("cache-control", "public, max-age=86400");
+      // Let the CDN and the browser cache for 24hrs. specify ?nocache to bypass caching.
```
Comment (mikelehen) [inline functions/src/index.ts]: nocache => "no-cache" ?  Also, is this just for testing?  Maybe Make the comment say, "For testing you can specify ?no-cache to bypass caching." or whatever. 🤷

File: functions/src/index.ts [hunk 7]
```diff
@@ -145,9 +164,11 @@ app.get("/screenshot", (req, res) => {
...
+      if (req.query["no-cache"] === undefined) {
+        console.log("Setting cache-control header.");
```
Comment (mikelehen) [inline functions/src/index.ts]: intentional console.log?

**Review by mikelehen — APPROVED** [review]
Looks good with some minor feedback.

File: README.md (hunk 4 above)
Comment (smcclure17) [reply README.md]: ah, this is the `Required` field, but it's hard to tell as the formatting was all skewed

File: functions/src/index.ts (hunk 3 above)
Comment (smcclure17) [reply functions/src/index.ts]: >  think you can get rid of as "2GB" if you do const `runtimeOpts: functions.RuntimeOptions = { ..`

Ahh, nice--thank you! and good call, added a comment

**Pull request merged**